crypto/mem_sec.c

```diff
@@ -588,6 +588,9 @@ static void *sh_malloc(size_t size)
 
     OPENSSL_assert(WITHIN_ARENA(chunk));
 
+    /* zero the free list header as a precaution against information leakage */
+    memset(chunk, 0, sizeof(SH_LIST));
+
```

Comment:
Would it make sense to make this a `CLEAR(ret, actual_size);` up in `CRYPTO_secure_malloc` instead, just as there is one in `CRYPTO_secure_free`?

Reply:
Since `CRYPTO_secure_malloc()` now zeroes memory by default, it makes sense to also change `CRYPTO_secure_zalloc()` as follows in order to avoid a double memset:

```C
void *CRYPTO_secure_zalloc(size_t num, const char *file, int line)
{
   if (!secure_mem_initialized) {
        return CRYPTO_zalloc(num, file, line);
   }
   /* CRYPTO_secure_malloc() now zeroes memory by default */
   return CRYPTO_secure_malloc(num, file, line);
}
```

Reply:
Or better, call `CRYPTO_zalloc()` from `CRYPTO_secure_malloc() ` also in the `!secure_mem_initialized` case and make those two functions identical:

```
void *CRYPTO_secure_zalloc(size_t num, const char *file, int line)
{
   /* CRYPTO_secure_malloc() now zeroes memory by default */
   return CRYPTO_secure_malloc(num, file, line);
}
```

Reply:
Oh, I overlooked that @paulidale's original solution erased only `sizeof(SH_LIST)` bytes, not the entire allocated buffer. Wouldn't it be better to just leave it like this instead of moving the memset on level up, @levitte?

Reply:
The buffer will be cleared on `free` and is cleared after the initial `mmap(2)`.  Thus, it will always be zero unless it is allocated (once the free list header is dealt with).  So, there should be no need to clear the entire block of memory before returning as @mspncp pointed out.

I do think it makes sense to alias `CRYPTO_secure_zalloc` to `CRYPTO_secure_malloc`.  I'll double check and update in the morning (twelve or so hours).

@levitte, the free list structure contains two pointers, using CLEAR seems like overkill.  Most compilers will optimise the memset away in favour of a store or two.  As for moving it up a level, I'd be happy to do that but I think keeping it in the lower level routine makes more sense since it is dealing with the free list header struct which the higher levels have no need to know about.